// chractor/app/data.ts
module app {
  export interface IData {
        throttle: 0,
        slip : 0,
        gear : 0,
        driveshaft_rpm : 0,
        trans_temp" : 0,
        engine_temp : 0,
        tc_temp : 0,
        rad_temp : 0,
        air_temp : 0,
        throttle : 0,
        battery : 0,
        locked : 0
  }

  function getParameterByName(name: string, url?: string) {
    if (!url) url = window.location.search;
    name = name.replace(/[\[\]]/g, "\\$&");
    const results = new RegExp("[?&]" + name + "(=([^&#]*)|&|#|$)").exec(url);
    if (!results) return null;
    if (!results[2]) return '';
    return decodeURIComponent(results[2].replace(/\+/g, " "));
  }

  export function Fetch() {
    const fileName = "http://192.168.1.110:8080/api";
    console.log(`Starting fetch for ${fileName}`);

    FetchInternal(fileName);
  }

  export function FetchInternal(fileName: string) {
    const request = new XMLHttpRequest();

    request.addEventListener("load", e => TransferComplete(e, fileName));
    request.addEventListener("error", e => TransferFailed(e, fileName));
    request.addEventListener("abort", e => TransferCanceled(e, fileName));

    request.open("GET", fileName);
    request.send();
  }

  function TransferComplete(evt, fileName: string) {
    const result: IData = JSON.parse(evt.target.responseText);
    app.DisplayData(result);

    setTimeout(() => {
      app.FetchInternal(fileName);
    }, 500);
  }

  function TransferFailed(evt, fileName: string) {
    console.error("An error occurred while transferring the file.", evt);
    setTimeout(() => {
      app.FetchInternal(fileName);
    }, 2500);
  }

  function TransferCanceled(evt, fileName: string) {
    console.error("The transfer has been canceled by the user.", evt);
  }
}
