// chractor/app/display.ts
module app {
  export function DisplayData(data: IData): void {
    SetPercentage(
      "#slip",
      data.slip,
      {
        Good: { Min: 0, Max: 20 },
        Warning: { Min: 20, Max: 40 },
        Bad: { Min: 40, Max: 100 }
      });
    SetPercentage("#throttle", data.throttle);
    SetNumber(
      "#oil-pressure",
      data.oil,
      0,
      {
        Good: { Min: 0, Max: 50 },
        Bad: { Min: 50, Max: 1000 }
      });
    SetBatteryVoltage("#battery-voltage", data.battery);
    SetValue("#differential", data.lock ? "ON" : "OFF", data.lock ? "warning" : "");
    SetNumber("#gear", data.gear, 0);
    SetPercentage("#belt_slip", data.slip);
  }

  interface Interval {
    Min: number;
    Max: number;
  }

  interface Ranges {
    Good?: Interval;
    Warning?: Interval;
    Bad?: Interval;
  }

  function RangeClassFromRanges(value: number, ranges?: Ranges) {
    if (!ranges) {
      return "";
    }

    if (ranges.Good && ranges.Good.Min <= value && ranges.Good.Max >= value) {
      return "good";
    }

    if (ranges.Warning && ranges.Warning.Min <= value && ranges.Warning.Max >= value) {
      return "warning";
    }

    if (ranges.Bad && ranges.Bad.Min <= value && ranges.Bad.Max >= value) {
      return "bad";
    }
  }

  function SetPercentage(selector: string, value: number, ranges?: Ranges): void {
    SetValue(selector, value.toString() + " %", RangeClassFromRanges(value, ranges));
  }

  function SetDegrees(selector: string, value: number, ranges?: Ranges): void {
    SetValue(selector, value.toFixed(1).toString() + " °C", RangeClassFromRanges(value, ranges));
  }

  function SetBallast(selector: string, value: string): void {
    var stringValue = "OFF";
    switch (value) {
      case "Forward":
        stringValue = "FWD";
        break;
      case "Backward":
        stringValue = "REV";
        break;
    }
    SetValue(selector, stringValue);
  }

  function SetCVT(selector: string, value: number): void {
    if (value == -1) {
      SetValue(selector, "-");
    }
    else {
      SetNumber(selector, value);
    }
  }

  function SetBatteryVoltage(selector: string, value: number): void {
    var rangeClass = RangeClassFromRanges(
      value,
      {
        Good: { Min: 11, Max: 12.999999 },
        Warning: { Min: 13, Max: 15 },
        Bad: { Min: 15.000001, Max: 1000 }
      })
      if (value < 11) {
        rangeClass = "warning";
      }
    SetValue(selector, value.toFixed(1).toString(), rangeClass);
  }

  function SetNumber(selector: string, value: number, precision: number = 2, ranges?: Ranges): void {
    SetValue(selector, value.toFixed(precision).toString(), RangeClassFromRanges(value, ranges));
  }

  function SetValue(selector: string, value: string, rangeClass?: string): void {
    var root = <HTMLElement>document.querySelector(selector);
    root.classList.remove("good", "warning", "bad");
    if (rangeClass) {
      root.classList.add(rangeClass);
    }
    var element = <HTMLElement>root.querySelector(".value");
    element.innerHTML = value.toString();
  }
}
